### PR TITLE
FEATURE: Add node inline operation lifecycle

### PR DESCRIPTION
**Pull request by grebaldi**
This pull request will contain the entire lifecycle of adding a node in the guest frame.

TODOS:

- [x] Extend add node mechanism to pass relevant fusion paths to the server
- [x] Render the created node on server side and give feedback
- [x] Insert the rendered node at the right place in the guest frame DOM
- [x] Add new node to the store

### CODE REVIEW

File: Classes/Neos/Neos/Ui/Domain/Model/Feedback/Operations/RenderNode.php [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        return $this->node;
+    }
+
+    /**
+     * Set the referenceData
+     *
+     * @param string $referenceData
+     * @return void
+     */
+    public function setReferenceData(array $referenceData)
```
Comment (skurfuerst) [inline Classes/Neos/Neos/Ui/Domain/Model/Feedback/Operations/RenderNode.php]: instead of get/setReferenceData, I'd rather like to use "typoScriptPathToBeRendered" or so

File: Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php [hunk 2]
```diff
@@ -20,6 +22,13 @@
     protected $nodeType;
 
     /**
+     * The referenceData for the rendered node
```
Comment (skurfuerst) [inline Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php]: Actually, I'd even more prefer if we could not use an array here, but more specifically set the Fusion path as string explicitely

File: packages/neos-ui-redux-store/src/CR/Nodes/index.js [hunk 3]
```diff
@@ -95,11 +94,10 @@ export const reducer = handleActions({
             siteNode: $get('cr.nodes.siteNode', state) || '',
             focused: new Map({
                 contextPath: '',
-                typoscriptPath: ''
+                fusionPath: ''
```
Comment (skurfuerst) [inline packages/neos-ui-redux-store/src/CR/Nodes/index.js]: add a comment here why fusionPath is needed (e.g. only for out-of-band-rendering when adding new nodes)

File: packages/neos-ui-redux-store/src/UI/AddNodeModal/index.js [hunk 4]
```diff
@@ -23,7 +23,7 @@ export const actionTypes = {
 /**
  * Opens the add node modal.
  */
-const open = createAction(OPEN, (contextPath, mode) => ({contextPath, mode}));
+const open = createAction(OPEN, (reference, mode) => ({reference, mode}));
```
Comment (skurfuerst) [inline packages/neos-ui-redux-store/src/UI/AddNodeModal/index.js]: please add docs what "reference" is. String? Node Context Path? Object?

File: packages/neos-ui-redux-store/src/UI/AddNodeModal/index.js [hunk 5]
```diff
@@ -59,25 +60,32 @@ export const reducer = handleActions({
     [system.INIT]: () => $set(
         'ui.addNodeModal',
         new Map({
-            referenceNode: '',
+            reference: null,
```
Comment (skurfuerst) [inline packages/neos-ui-redux-store/src/UI/AddNodeModal/index.js]: hm, what is "reference"? I find it's hard to understand, because it is unclear what is in here and why we need it.

File: packages/neos-ui/src/Containers/ContentCanvas/InlineUI/NodeToolbar/Buttons/AddNode/index.js [hunk 6]
```diff
@@ -37,9 +39,23 @@ export default class AddNode extends PureComponent {
...
         const {
             openAddNodeModal,
-            node
+            fusionPath,
+            contextPath
         } = this.props;
+        const closestCollectionElement = dom.closestNode(
+            dom.findNode(contextPath, fusionPath)
+        );
+        const collection = {
```
Comment (skurfuerst) [inline packages/neos-ui/src/Containers/ContentCanvas/InlineUI/NodeToolbar/Buttons/AddNode/index.js]: collection -> "contentCollection". becuase that's what it is.

File: packages/neos-ui/src/Containers/ContentCanvas/InlineUI/NodeToolbar/Buttons/AddNode/index.js [hunk 7]
```diff
@@ -37,9 +39,23 @@ export default class AddNode extends PureComponent {
     handleOpenModalBtnClick() {
         const {
             openAddNodeModal,
-            node
+            fusionPath,
+            contextPath
         } = this.props;
+        const closestCollectionElement = dom.closestNode(
```
Comment (skurfuerst) [inline packages/neos-ui/src/Containers/ContentCanvas/InlineUI/NodeToolbar/Buttons/AddNode/index.js]: add a TODO here, that we always assume that closestNode is a collection; and that is kinda fragile I think...

File: packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js [hunk 8]
```diff
@@ -12,26 +12,29 @@ import style from './style.css';
 
 @connect($transform({
     hoveredNode: selectors.CR.Nodes.hoveredSelector,
-    focusedNode: selectors.CR.Nodes.focusedSelector
+    focused: $get('cr.nodes.focused')
```
Comment (skurfuerst) [inline packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js]: ? why is this change actually needed?

File: packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js [hunk 9]
```diff
@@ -12,26 +12,29 @@ import style from './style.css';
...
+            fusionPath: PropTypes.string
+        })
     };
 
     render() {
-        const {focusedNode, hoveredNode} = this.props;
+        const {focused, hoveredNode} = this.props;
 
-        if (!focusedNode && !hoveredNode) {
+        if (!focused.get('contextPath') && !hoveredNode) {
```
Comment (skurfuerst) [inline packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js]: use $get instead of assuming this is an immutable Object

File: packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js [hunk 10]
```diff
@@ -12,26 +12,29 @@ import style from './style.css';
...
 @connect($transform({
     hoveredNode: selectors.CR.Nodes.hoveredSelector,
-    focusedNode: selectors.CR.Nodes.focusedSelector
+    focused: $get('cr.nodes.focused')
 }))
 export default class InlineUI extends PureComponent {
     static propTypes = {
         hoveredNode: PropTypes.object,
-        focusedNode: PropTypes.object
+        focused: PropTypes.shape({
```
Comment (skurfuerst) [inline packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js]: probably this fails because the object is an immutable object?

File: packages/neos-ui/src/Containers/Modals/AddNode/index.js [hunk 11]
```diff
@@ -53,6 +54,16 @@ const calculateActiveMode = (currentMode, allowedNodeTypesByMode) => {
 }))
 export default class AddNodeModal extends PureComponent {
     static propTypes = {
+        reference: PropTypes.shape({
+            subject: PropTypes.shape({
+                contextPath: PropTypes.string.isRequired,
+                fusionPath: PropTypes.string
+            }).isRequired,
+            collection: PropTypes.shape({
```
Comment (skurfuerst) [inline packages/neos-ui/src/Containers/Modals/AddNode/index.js]: naming: what is "collection"? "enclosingContentCollection"?

File: Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php [hunk 12]
```diff
@@ -176,6 +213,20 @@ protected function createNode(NodeInterface $parent)
             $node->setProperty($key, $value);
         }
 
+        if ($nodeType->isOfType('TYPO3.Neos:Content') && $this->getReferenceData()['subject']['fusionPath']) {
+            $renderNode = new RenderNode();
```
Comment (skurfuerst) [inline Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php]: ...AddToContentCollectionOutOfBand is maybe even better; as the code in JS makes assumptions this is only working in Content Collections

File: Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php [hunk 13]
```diff
@@ -176,6 +213,20 @@ protected function createNode(NodeInterface $parent)
             $node->setProperty($key, $value);
         }
 
+        if ($nodeType->isOfType('TYPO3.Neos:Content') && $this->getReferenceData()['subject']['fusionPath']) {
```
Comment (skurfuerst) [inline Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php]: actually, I think this condition is not fully correct:

1) additionally, the parent node must *always* be a ContentCollection for having this feedback work.

2) if the parent is NO contentCollection, we should reload the page using the feedback we already have.

File: Classes/Neos/Neos/Ui/Domain/Model/Feedback/Operations/RenderNode.php (hunk 1 above)
Comment (dimaip) [inline Classes/Neos/Neos/Ui/Domain/Model/Feedback/Operations/RenderNode.php]: yes, but let's use the name fusion please

File: Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php (hunk 2 above)
Comment (grebaldi) [inline Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php]: This doesn't suffice. For inline node operations, we need to distribute fusion path for both, the reference node and the closest collection node, so the Feedback Handler knows exactly where to put the rendered node.

File: Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php (hunk 13 above)
Comment (grebaldi) [inline Classes/Neos/Neos/Ui/Domain/Model/Changes/AbstractCreate.php]: Is there actually any scenario, in which we create content nodes outside of content collections?

File: packages/neos-ui/src/Containers/ContentCanvas/InlineUI/NodeToolbar/Buttons/AddNode/index.js (hunk 7 above)
Comment (grebaldi) [inline packages/neos-ui/src/Containers/ContentCanvas/InlineUI/NodeToolbar/Buttons/AddNode/index.js]: See my comment above. I think this actually pretty implicit.

File: packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js (hunk 8 above)
Comment (grebaldi) [inline packages/neos-ui/src/Containers/ContentCanvas/InlineUI/index.js]: `focusedSelector` returns a node, `$get('cr.nodes.focused')` returns a contextPath and a fusionPath, which are the two values needed to safely identify a rendered node in the DOM.

File: packages/neos-ui/src/Containers/Modals/AddNode/index.js (hunk 11 above)
Comment (grebaldi) [inline packages/neos-ui/src/Containers/Modals/AddNode/index.js]: I'd prefer `closestContentCollection`, but I don't like bloating this, since it can't be anything else in this context. I'd rather change the name of `reference` to make more clear what it is about. Maybe something like `creationContext`.
